allure-kotlin-android/src/main/java/io/qameta/allure/android/runners/AllureAndroidJUnitRunners.kt

```diff
@@ -0,0 +1,98 @@
+package io.qameta.allure.android.runners
+
+import android.os.Bundle
+import androidx.multidex.MultiDex
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import androidx.test.runner.AndroidJUnitRunner
+import io.qameta.allure.android.AllureAndroidLifecycle
+import io.qameta.allure.android.internal.isDeviceTest
+import io.qameta.allure.android.internal.requestExternalStoragePermissions
+import io.qameta.allure.kotlin.Allure
+import io.qameta.allure.kotlin.junit4.AllureJunit4
+import org.junit.runner.*
+import org.junit.runner.manipulation.*
+import org.junit.runner.notification.*
+
+/**
+ * Wrapper over [AndroidJUnit4] that attaches the [AllureJunit4] listener and
+ * grants external storage permissions for tests running on a device (required for the test results to be saved).
+ */
+class AllureAndroidJUnit4(clazz: Class<*>) : Runner(), Filterable, Sortable {
```

Comment:
Monolith runner looks very complex. What do you think? Maybe we will implement separate runner for robolectric. In that case, we are able to change runners very simple if there appeared new one.

Reply:
Yep, seems like another open class for Robolectric runner will be better for intergration of other tools someone may use with a simple override impl.

Reply:
> Monolith runner looks very complex. What do you think? Maybe we will implement separate runner for robolectric. In that case, we are able to change runners very simple if there appeared new one.

I would argue it's not very complex. There is more documentation in this class than actual code :P Moreover this is the approach that Google is taking. They deprecated the old `AndroidJUnit4` runner and defined a centralized one that delegates the work to an appropriate class runner. Thanks to that you can annotate test class with a single runner & use the same code in both Robolectric and instrumentation tests (**sharedTest** in the sample). Otherwise, you are not able to do that. The actual runner is the implementation details that the client doesn't need to know about. 

> Yep, seems like another open class for Robolectric runner will be better for intergration of other tools someone may use with a simple override impl.

Do you have any example of another class runner? Android tests runners are quite complex and the only 2 supported ones that I know of are the instrumentation and Robolectric runner. I took the same approach as Google in their runner. If ever the need comes for more runners we can review it.